package.json

```diff
@@ -246,6 +246,7 @@
     "babel-eslint": "8.2.3",
     "chai": "4.1.2",
     "chai-enzyme": "1.0.0-beta.1",
+    "check-node-version": "3.2.0",
```

Comment:
No need to include this if we're using `npx` anyway.

Reply:
yeah, i'd pulled it out originally, but if we're going to be running it continually, seemed best to include it. We do get into a dependency pickle though unless we spec the version when we run it with npx.

Reply:
`npx` will install the package globally if needed, but after that time it will just use it because it's already installed. It doesn't download it every time, so I think your initial hunch to pull it out was better :)

Reply:
me too, though running `npx check-node-version --package` repeatedly, I'm seeing it reinstall it over and over and over... hrm.

Reply:
author of npx's take at https://github.com/zkat/npx/issues/154#issuecomment-379595899

> If you wanna specify a version, and you wanna use that version a lot, install it as a devDependency. That's the workflow it's designed for.

> I'm not going to cache things like npx foo@1.2.3 because I don't want to have to manage yet another global caching scheme and figuring out how to invalidate things. As far as npx is concerned, if it has to install something, then it's going to throw it away when it's done so it doesn't start filling your system up with crud -- that is, in fact, one of the main features of it, vs global installs.

> If you have a run-script, you have a package.json. If you have a package.json you have almost definitely done an npm install. If you have done an npm install, you can have devDependencies. If you're relying on npm install --only=prod, then install it as a regular dependency.

Reply:
For what I understood, `npx foo@1.2.3` isn't cached, but `npx foo` is. [Here](https://www.npmjs.com/package/npx) it says that it uses `npm`'s local cache by default. The first time I run `npx check-node-version` in my machine took ~5s, the next times took less than 2, so there's at least some caching involved. I would understand if you prefer keeping it in `devDependencies` though.

Reply:
I'm trying to understand what the tradeoffs are by not specifying it as `devDepencency`.

Semantically, this _is_ a dev dependency. It's a part of the development flow.

Technically, excluding it from `devDependencies` would exclude if from `package-lock`, `npm ci`, Docker builds, etc. Is that accurate?

`npx` gives us the capability to "depend" on things without declaring them. Install on demand.

If this is all accurate, it does seem like a small win to remove the declared dependency and let `npx` handle it.

Reply:
> Semantically, this is a dev dependency. It's a part of the development flow.

True. Adding to that, it's a "special" dependency in the sense that it's needed as a "pre-check", even if `node_modules/` doesn't exist. On other commands that can be executed without `node_modules/`, such as `bin/rm.js`, we took the approach of "use just the `node.js` built-in library". That could be another way to approach this, but `npx` seems to be better.

> Technically, excluding it from devDependencies would exclude if from package-lock, npm ci, Docker builds, etc. Is that accurate?

Also true.

> npx gives us the capability to "depend" on things without declaring them. Install on demand.

Yes. The first time it's installed "on the fly", and the next times it's cached so it's faster. *But*, for some reason the caching is not perfect, it still takes a couple of seconds to run `npx check-node-version`. If the package is in `node_modules/`, `npx` is instant.

> If this is all accurate, it does seem like a small win to remove the declared dependency and let npx handle it.

As you said before, it's a trade-off. Oh-so-slightly longer `npm ci` + slightly bigger `package-lock.json` vs making each `npm start` a couple of seconds faster.